Improvements to location columns in app status report

mkangia opened the pull request:
## Product Description
<!-- Where applicable, describe user-facing effects and include screenshots. -->
In order to prep for releasing as GA, Some improvements to the feature flag that adds location columns to app status report (now known as "User Last Activity Report")
- Move location columns to the end of the excel file keeping in mind domains that might have too many location columns
- default to including all location types in the sheet instead of just the ones that are needed for code benefit
- quite some code improvements and test addition

## Technical Summary
<!--
    Provide a link to any tickets, design documents, and/or technical specifications
    associated with this change. Describe the rationale and design decisions.
-->
https://dimagi.atlassian.net/browse/SAAS-18743
:fish: review by commit

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
`LOCATION_COLUMNS_APP_STATUS_REPORT`

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
Tested locally, added tests

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
QA request submitted for regression testing
https://dimagi.atlassian.net/browse/QA-8213

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [X] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change

mkangia committed:
navigate code

- add helpful comments
- cleanup: fix incorrect param list

mkangia committed:
improve code structure

- reorder methods
- make methods public
- memoize repeated use

mkangia committed:
variable name change

mkangia committed:
better naming

- _get_location_columns to _get_relevant_location_types
- loc_col to loc_type

mkangia committed:
better naming

- _include_location_data to _include_ancestor_locations_data

mkangia committed:
better naming & scope

- _include_location_data to _include_ancestor_locations_data
- relevant_location_types to _relevant_location_types

mkangia committed:
default to all location types

mkangia committed:
avoid passing location types

mkangia committed:
rename and move method

mkangia committed:
move location columns to the end of the file

mkangia committed:
skip fetching ancestors when no locations

- an empty set of locations leads to query failure when fetching ancestors

mkangia committed:
translate

mkangia committed:
add tests for export table

- add more users
- add diff app version for users
- assign diff location to diff users

mkangia committed:
refactor: better names and documentation

millerdev commented on corehq/apps/reports/tests/test_application_status_report.py:
```diff
@@ -42,15 +51,47 @@ def setUpClass(cls):
+        loc_city.save()
+
+        # cleans up the full hierarchy
+        cls.addClassCleanup(loc_country.location_type.delete)
```
Are these in SQL? If yes, they will be cleaned up automatically by transaction rollback in class cleanup.

mkangia committed:
review feedback

- fix/update documentation
- add hint for translation
- remove redundant cleanup in test

millerdev commented on corehq/apps/reports/standard/deployments.py:
```diff
@@ -408,14 +353,66 @@ def process_rows(self, users, fmt_for_export=False):
+        Returns the hierarchy for locations
+        {
+            location_id_1: [self, parent, parent_of_parent,...],
+            location_id_2: [self, parent, parent_of_parent,...],
```
nit: missing space
```suggestion
            location_id_1: [self, parent, parent_of_parent, ...],
            location_id_2: [self, parent, parent_of_parent, ...],
```

mkangia committed:
add missing space

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

mkangia commented:
Thanks for the review @millerdev 
Need an approval post the last suggestion committed

kaapstorm reviewed (approved):
Two suggestions (although I strongly recommend using `deepcopy()`). Nothing blocking.

mkangia committed:
cache hierarchy

mkangia committed:
better naming

mkangia committed:
limit just to pk

mkangia committed:
limit location column to title
